Mathlib/Data/Multiset/DershowitzManna.lean

```diff
@@ -0,0 +1,409 @@
+
+## References
+
+* [Wikipedia, Dershowitz–Manna ordering*]
+(https://en.wikipedia.org/wiki/Dershowitz%E2%80%93Manna_ordering)
+
+* [CoLoR](https://github.com/fblanqui/color), a Coq library on rewriting theory and termination.
+  Our code here is inspired by their formalization and the theorem is called `mOrd_wf` in the file
+  [MultisetList.v](https://github.com/fblanqui/color/blob/1.8.5/Util/Multiset/MultisetOrder.v).
+
+-/
+
+variable {α : Type*}
+
+/-- The standard Dershowitz–Manna ordering: -/
+inductive MultisetDMLT [DecidableEq α] [Preorder α] (M N : Multiset α) : Prop :=
+  | DMLT : ∀ (X Y Z : Multiset α),
+    Y ≠ ∅ →
+    M = Z + X →
+    N = Z + Y →
+    (∀ x ∈ X, ∃ y ∈ Y, x < y) → MultisetDMLT M N
+
+/-- MultisetRedLT is a special case of MultisetDMLT. The transitive closure of it is used to define
+    an equivalent (proved later) version of the ordering. -/
+inductive MultisetRedLT [DecidableEq α] [LT α] (M N : Multiset α) : Prop :=
+  | RedLT : ∀ (X Y : Multiset α) (a : α) ,
+    (M = X + Y) →
+    (N = X + {a}) →
+    (∀ y, y ∈ Y → y < a) → MultisetRedLT M N
+
+open Relation
+
+/-- MultisetTransLT is the transitive closure of MultisetRedLT and is equivalent to MultisetDMLT
+    (proved later). -/
+def MultisetTransLT [DecidableEq α] [LT α] : Multiset α → Multiset α → Prop :=
+    TransGen MultisetRedLT
+
+/-- A shorthand notation: AccM defines the accessibility relation given MultisetRedLT. -/
+def AccM [DecidableEq α] [Preorder α] : Multiset α → Prop := Acc MultisetRedLT
+
+/- MultisetRedLT is a special case of MultisetDMLT. -/
+theorem dmLT_of_redLT [DecidableEq α] [Preorder α] (M N : Multiset α) (h : MultisetRedLT M N):
+    MultisetDMLT M N := by
+  rcases h with ⟨X, Y, a, M_def, N_def, ys_lt_a⟩
+  apply MultisetDMLT.DMLT Y {a} X _ M_def N_def
+  · simpa
+  · simp
+
+/- Some useful lemmas. -/
+lemma not_redLT_zero [DecidableEq α] [LT α] (M: Multiset α) : ¬ MultisetRedLT M 0 := by
+  intro h
+  cases h with
+  | RedLT X Y a M nonsense _ =>
+    have contra : a ∈ (0 : Multiset α):= by
+      rw [nonsense]
+      simp_all only [Multiset.mem_add, Multiset.mem_singleton, or_true]
+    contradiction
+
+lemma red_insert [DecidableEq α] [LT α] {a : α} {M N : Multiset α} (h : MultisetRedLT N (a ::ₘ M)) :
+    ∃ M',
+      N = (a ::ₘ M') ∧ (MultisetRedLT M' M) ∨ (N = M + M') ∧ (∀ x : α, x ∈ M' → x < a) := by
+  rcases h with ⟨X, Y, a0, h1, h0, h2⟩
+  by_cases hyp : a = a0
+  · exists Y; right; apply And.intro
+    · rw [h1, add_left_inj]
+      rw [add_comm, Multiset.singleton_add] at h0
+      simp_all
+    · simp_all
+  · exists (Y + (M - {a0}))
+    left
+    constructor
+    · rw [h1]
+      have : X = (M - {a0} + {a}) := by
+        rw [add_comm, Multiset.singleton_add] at *
+        ext b
+        simp [Multiset.ext, Multiset.count_cons] at h0
+        by_cases h : b = a
+        · have := h0 b
+          aesop
+        · have := h0 b
+          simp [Multiset.sub_singleton]
+          aesop
```

Comment:
another instance of an `aesop` call that may be replaced with the result of `aesop?` (I think)